compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/IrSourceCompilerForInline.kt

```diff
@@ -95,31 +87,14 @@ class IrSourceCompilerForInline(
         }
 
     override val lazySourceMapper: DefaultSourceMapper
-        get() = codegen.classCodegen.getOrCreateSourceMapper()
-
-    private fun makeInlineNode(function: IrFunction, classCodegen: ClassCodegen, isLambda: Boolean): SMAPAndMethodNode {
-        var node: MethodNode? = null
-        val functionCodegen = object : FunctionCodegen(function, classCodegen, codegen.takeIf { isLambda }) {
-            override fun createMethod(flags: Int, signature: JvmMethodGenericSignature): MethodVisitor {
-                val asmMethod = signature.asmMethod
-                node = MethodNode(
-                    Opcodes.API_VERSION,
-                    flags,
-                    asmMethod.name.removeSuffix(FOR_INLINE_SUFFIX),
-                    asmMethod.descriptor,
-                    signature.genericsSignature,
-                    null
-                )
-                return wrapWithMaxLocalCalc(node!!)
-            }
+        get() = codegen.smapOverride ?: codegen.classCodegen.sourceMapper.also {
```

Comment:
Wouldn't be here some kind of instability in case for example of IC?

Reply:
You mean something like
```
// a.kt
fun f() = h()
// b.kt
fun g() = otherInlineFun() // written to source map first if b.kt is compiled separately
inline fun h() = otherInlineFun() // written to source map first if a.kt and b.kt are compiled together
// c.kt
inline fun otherInlineFun() {}
```
? It actually would, yes. Ideally, writes to the source map should be done in source order, just like method nodes are written into methods by ClassCodegen and not whenever IrSourceCompilerForInline requests them.

Of course, the current state isn't free from instability either:
```
// a.kt
fun f() = h()
// b.kt
// These two methods reference different classes if compiled separately:
fun g() = h()
inline fun h() = myRun { "OK" }
// c.kt
inline fun <T> myRun(crossinline x: () -> T) = { x() }()
```

Reply:
One way to solve this that I see is to create separate source maps for each method and cache them together with method nodes, then in `generateMethod` merge each cached SMAP into the class and use a method visitor to remap the line numbers.

Reply:
@pyos Could you also add IC test (@altsvetkov Please help with best form for it)

Reply:
> One way to solve this that I see is to create separate source maps for each method and cache them together with method nodes, then in generateMethod merge each cached SMAP into the class and use a method visitor to remap the line numbers.

If there independent part in this pr it could be extracted in separate one

Reply:
Extracted various refactorings into #3205 - the rest of the changes are not as big: https://github.com/pyos/kotlin/compare/ir/refactor-class-codegen..ir/reuse-inline-methods
I have a general idea on how to make source maps stable, but this will require modifying SMAP-related classes, so I'll submit that part separately as well.

Reply:
Last moment thought: @altsvetkov Is it correct that if 'b.kt' is changed somehow but `h` and `myRun` untouched then 'a.kt' wouldn't recompiled by IC. Or it's would be recompiled in anyway cause of inline dependency?

Reply:
That's actually a good question - if the line numbers in b.kt shift, then the source maps in AKt need to be updated even if `h` stayed the same otherwise...

Reply:
NB: just understand that changed smap in 'b' implies different linenumbers in 'h' (cause there would be another fake line numbers). So 'a' also would be affected cause of changed bytecode of inline function (looks like that fake linenumbers in 'a' would be same except also changed smap mapping). So recompilation order in IC would looks like b+c, then a (@altsvetkov am I correct? Would be any additional IC phase triggered?)

Reply:
> Is it correct that if 'b.kt' is changed somehow but `h` and `myRun` untouched then 'a.kt' wouldn't recompiled by IC. Or it's would be recompiled in anyway cause of inline dependency?

It will be recompiled, if bytecode is changed. We estimate changes in bytecode by calculating MD5 hash of a JVM method corresponding to an inline declaration. Currently usages of modified inline function are recompiled separately during the next round, but we may change that behaviour in future (currently it is possible to incrementally define a cycle of inline functions (which would be impossible to compile together, but can be compiled separately)). Here is the code calculating hashes https://github.com/JetBrains/kotlin/blob/master/build-common/src/org/jetbrains/kotlin/incremental/IncrementalJvmCache.kt#L491

Reply:
> NB: just understand that changed smap in 'b' implies different linenumbers in 'h' (cause there would be another fake line numbers). So 'a' also would be affected cause of changed bytecode of inline function (looks like that fake linenumbers in 'a' would be same except also changed smap mapping. Question what would be with 'a'when 'h' bytecode is totatlly same are still open). So recompilation order in IC would looks like b+c, then a (@altsvetkov am I correct? Would be any additional IC phase triggered?)

First lets check, if I understand dependencies correctly: a.kt depends on b.kt, which in turn depends on c.kt (depends == by inlining). 
If c.kt is changed, then IC will do the following:
1. Compile c.kt, compare proto + bytecode of `otherInlineFun`.
2. If bytecode of `otherInlineFun` is changed, then compile and compare `b.kt`.
3. If bytecode of `h` is changed, then compile and compare `a.kt`.

The algorithm is the same:
0. All source files modified by user are considered dirty.
1. Dirty files are compiled, resulting proto and bytecode for inline declarations/constants are compared with cached values.
2. Dirty set is cleared.
3. Usages of modified FQ-names are added to dirty set (by searching lookups).
4. If new dirty set is not empty, go to 1.

Reply:
For tests check test data in https://github.com/JetBrains/kotlin/tree/master/jps-plugin/testData/incremental/incrementalJvmCompilerOnly

All test are a bunch of source files and steps. Each step creates, modifies, deletes or "touches" some of source files (similar to bunch files `a.kt.new.1` means that on step 1 `a.kt` content will be replaced with the content of `a.kt.new.1`; instead of `new` you can use `delete` or touch where appropriate). Conceptually after each step we run IC and compare what files were compiled with expected files (in actuality we log results of each step, and compare all steps with build.log.txt in the end). If all steps performed IC as expected (i.e. only expected files were compiled on each step), then we rebuild all files in a test case, and compare incremental and non-incremental outputs.

Reply:
@altsvetkov thanks for the explanation.

@max-kammerer I've fixed this problem and added two incremental compilation tests, both of which pass with the PR - a strict improvement over the current situation (one of them also fails on the non-IR backend). `inlineFunctionRegeneratedObjectStability` could still be modified to fail by having multiple inline functions with the same name, though. In general, it'd probably be best to refactor ClassCodegen to have an object wrapping the class visitor/smap/etc. with `generateField`/`generateMethod` as extension methods on that object, and avoid giving a reference to this object to `FunctionCodegen` - this would completely prevent this sort of error, but such a refactoring is not trivial due to the fact that local classes can be anywhere within function bodies.